packages/vite/src/node/plugins/optimizedDeps.ts

```diff
@@ -0,0 +1,110 @@
+import { promises as fs } from 'fs'
+import type { Plugin } from '../plugin'
+import colors from 'picocolors'
+import { cleanUrl, createDebugger } from '../utils'
+import { isOptimizedDepFile } from '../optimizer'
+import type { DepOptimizationMetadata, OptimizedDepInfo } from '../optimizer'
+import type { ViteDevServer } from '..'
+
+export const ERROR_CODE_OPTIMIZE_DEPS_TIMEOUT = 'ERR_OPTIMIZE_DEPS_TIMEOUT'
+export const ERROR_CODE_OPTIMIZE_DEPS_OUTDATED = 'ERR_OPTIMIZE_DEPS_OUTDATED'
+
+const isDebug = process.env.DEBUG
+const debug = createDebugger('vite:optimize-deps')
+
+export function optimizedDepsPlugin(): Plugin {
+  let server: ViteDevServer | undefined
+
+  return {
+    name: 'vite:optimized-deps',
+
+    configureServer(_server) {
+      server = _server
+    },
+
+    apply: 'serve',
+
+    async load(id) {
+      if (server && isOptimizedDepFile(id, server.config)) {
+        const metadata = server?._optimizeDepsMetadata
+        if (metadata) {
+          const file = cleanUrl(id)
+          const info = optimizeDepInfoFromFile(metadata, file)
+          if (info) {
+            try {
+              // This is an entry point, it may still not be bundled
+              await info.processing
+            } catch {
+              // If the refresh has not happened after timeout, Vite considers
+              // something unexpected has happened. In this case, Vite
+              // returns an empty response that will error.
+              optimizeDepsTimeout(id)
+              return
+            }
+            const newMetadata = server._optimizeDepsMetadata
+            if (metadata !== newMetadata) {
+              const currentInfo = optimizeDepInfoFromFile(newMetadata!, file)
+              if (info.browserHash !== currentInfo?.browserHash) {
+                outdatedTimeout(id)
+              }
+            }
+          }
+          isDebug && debug(`load ${colors.cyan(file)}`)
```

Comment:
I don't know, there are several other parts of the codebase using the `isDebug && debug(...)` pattern, maybe to avoid the call to colors? 🤔

Reply:
yes, this pattern is usually to avoid extra work in building the log string if the logger is disabled

Reply:
Makes sense. I doesn't seem like it's enforced everywhere though, and doesn't check for the debug namespace. But looks fine for now since there are existing patterns.